lib/fs/fakefs.go

```diff
@@ -163,9 +170,19 @@ func (fs *fakefs) entryForName(name string) *fakeEntry {
 			return nil
 		}
 		var ok bool
-		entry, ok = entry.children[comp]
-		if !ok {
-			return nil
+
+		if fs.insens {
+			entry, ok = entry.children[unicodeFoldLower(comp)]
+			if !ok {
+				return nil
+			}
+
+			entry.name = comp
```

Comment:
We might need to copy entry? Because this can change the value after it has been returned.

Reply:
I can't see where this happens (other than in places where there are already tests to prove things work as expected).

What I did find while looking for those, however, is DirNames that supposedly returns names for all the files in a directory. There are no tests for it, so I missed it initially. Do we use it? Because if we do and if we rely on it, I'd better revert back to my initial approach: I can't fathom a way to make such a function to return file names in their original capitalization if our index is folded :(

Reply:
Ok, judging by the tests it looks to me that everything does work as expected now. Ready for review.

Reply:
You don't have a test for this.

The test would be create aaa, stat AAA (you will get AAA) store the result, stat again AaA, check stored AAA, it would become AaA.

Reply:
I don't see why it would behave the way you described given how code is written, but I've just added a test to be sure it behaves as intended ;)

Reply:
The stat result is intentionally a copy, so won't reflect later changes:

https://github.com/syncthing/syncthing/blob/28ad9ba434c49e2b06964a2dbb4022ba470524e5/lib/fs/fakefs.go#L737-L740

but it looks like you're modifying the original, this changing the file name "on disk" by stat:ing it?

Reply:
Yeah you'd have to restat the first file handle.

Reply:
OK, I officially admit I'm too dumb for this. Reading and re-reading the code most carefully what you both are saying makes perfect sense, and this new version of the test should most definitely fail, because, as demonstrated by another test, DirNames does show what's actually in the directory.

Problem is, the test passes. :/

If you have time for this when you come back, I'd really appreciate you showing me where things go wrong. Otherwise, don't bother: wasting too much of maintainers' time is not something I'm aiming to do here.

Reply:
What exactly do you think should fail?  
Unless I misunderstood something (totally possible :) ) I do see the issue pointed out in this conversation and a few others (see rest of review comments), so I generally agree a failure would be nice/expected. However the nature of the issues (using the map keys or argument as names in the returned values) ensures always consistent (while incorrect) results thus tests don't fail. I don't find a test that triggers the particular failure pointed out here. For that you'd e.g. create `Foo`, stat `foo` and then check that what's "on disk" (in fakefs) is still `Foo` (by manually traversing the fake-entry-maps). However that's not even the case currently as on creation the name is already folded.

As I can't comment on `DirNames` (no change): You need to adapt it to not return the map keys as names, but the actual unfolded names of the map entries.